README.md
# C(ontinued)-MaNGOS -- README

This file is part of the CMaNGOS Project. See [AUTHORS](doc/AUTHORS.md) file for Copyright information

## Welcome to C(ontinued)-MaNGOS

C(ontinued)-MaNGOS is a free project with the following goal:

  **Doing WoW-Emulation Right!**

This means, we want to focus on:

* WoW-Emulation
    This project is about developing a server software that is able to
    emulate a well known MMORPG service.

* Doing
  * This project is focused on developing content!
  * Also there are many other aspects that need to be done and are
    considered equally important.
  * Anyone who wants to do stuff is very welcome to do so!

* Right
  * Our goal must always be to provide the best code that we can.
  * Being 'right' is defined by the behavior of the system
    we want to emulate.
  * Developing things right also includes documenting and discussing
    _how_ to do things better, hence...
  * Learning and Teaching are very important in our view, and must
    always be part of what we do.

To be able to accomplish these goals, we support and promote:

* Freedom
  * of our work: Our work - including our code - is released under the GPL
    So everbody is free to use and contribute to this open source project
  * for our developers and contributers on things that interest them.
    No one here is telling anybody _what_ to do.
    If you want somebody to do something for you, pay him,
    but we are here to enjoy.
  * to have FUN with developing.

* A friendly environment
  * We try to let personal issues behind us.
  * We only argue about content, and not about thin air!
  * We follow the [Netiquette](http://tools.ietf.org/html/rfc1855).

-- The C(ontinued)-MaNGOS Team!

## Further information

  You can find further information about CMaNGOS at the following places:
  * [CMaNGOS Forum](http://cmangos.net/)
  * CMaNGOS IRC channel: #cmangos on irc.rizon.net
  * [GitHub repositories](https://github.com/cmangos/)
  * [Issue tracker](https://github.com/cmangos/issues/issues)
  * [Pull Requests](https://github.com/cmangos/mangos-wotlk/pulls)
  * [Current build status on Travis CI](https://travis-ci.org/cmangos/mangos-wotlk/)
  * [Wiki](https://github.com/cmangos/issues/wiki) with additional information on installation
  * [Contributing Guidelines](doc/CONTRIBUTING.md)
  * Documentation can be found in the doc/ subdirectory and on the github wiki

## License

  CMaNGOS is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


  You can find the full license text in the file [COPYING](doc/COPYING.md) delivered with this package.

### Exceptions to GPL

  Some third-party libraries CMaNGOS uses have other licenses, that must be
  uphold.  These libraries are located within the dep/ directory

  In addition, as a special exception, the CMaNGOS project
  gives permission to link the code of its release of MaNGOS with the
  OpenSSL project's "OpenSSL" library (or with modified versions of it
  that use the same license as the "OpenSSL" library), and distribute
  the linked executables.  You must obey the GNU General Public License
  in all respects for all of the code used other than "OpenSSL".  If you
  modify this file, you may extend this exception to your version of the
  file, but you are not obligated to do so.  If you do not wish to do
  so, delete this exception statement from your version.


doc/NEWS.md
# CMaNGOS -- History of visible changes

This file is part of the CMaNGOS Project. See [AUTHORS](AUTHORS.md) file for Copyright information

Most of our code is released under the GPL v2, a copy is shipped with the file [COPYING](COPYING).

Visit our project website for forums, links and more:
http://cmangos.net

Please submit bug reports at our bugtracker at:
https://github.com/cmangos/issues

See the file [CHANGELOG](CHANGELOG.md) for details!

## Version 0.17
 * Move to CMaNGOS (http://cmangos.net - https://github.com/cmangos)
 * Add Prepared SQL-Statements
 * Updated MySQL client libs to 5.1.49.
 * Updated ACE framework to 5.8.3.
 * Updated G3DLite library to 8.0.1

## Version 0.16
 * G3DLite lib upgrade to G3DLite 8.0b4
 * OpenSSL lib upgrade to OpenSSL 1.0.0.
 * Sockets lib use dropped.
 * dotconfpp lib use dropped.
 * Implemented SOAP in MaNGOS.
 * Upgrade to client version 3.3.3a (build 11723).

## Version 0.15
 * Implement AV battleground.
 * Implement new DB-based gossip system for creatures/gameobjects.
 * Implement realmd support realms for different client versions.
 * New memory allocator for MaNGOS, based on Intel Threading Building Blocks library.
 * Upgrade to client version 3.2.2a (build 10505).

## Version 0.14
 * Implement possibility to reset client cache data from server side.
 * DB-based store battleground gameobjects/creatures.
 * Map search/loading improvements
 * Upgrade to client version 3.1.3 (build 9947).

## Version 0.13
 * OpenSSL lib upgrade to OpenSSL 0.9.8k.
 * ZThread replaced by similar ACE framework functionality. ZThread use dropped.
 * Visual Studio 2003 (VC71) support dropped.
 * Experemental support for Visual Studio 2010 Beta 1 added.
 * Sqlite sql engine support dropped.
 * Upgrade to client version 3.0.9 (build 9551).

## Version 0.12
 * Migrate from SVN to GIT.
 * Implemented arena system.
 * Start using Adaptive Communication Environment (ACE) framework v.5.6.6 in server network code.
 * Update Visual Leak Detector lib to Version 1.9g (beta).
 * Upgrade to client version 2.4.3 (build 8606).

## Version 0.11
 * Lots of improvements in the spell system.
 * Now use Mersenne Twister random number generator.
 * MySQL client libs updated to 5.0.56.
 * Upgrade to client version 2.4.2 (build 8278).

## Version 0.10
 * master looter lot mode,
 * FFA PvP zones and FFA PvP server type support,
 * Guild Banks,
 * unique equipped items support, including gems,
 * talent inspecting,
 * PostgreSql support,
 * sockets library updated to v.2.2.9 version,
 * 64-bit binaries building at Windows,
 * Upgrade to client version 2.3.0 (build 7561).

## Version 0.9
 * recipes discovery system support,
 * more auras, effects, spells, and talents working,
 * players queue at login support,
 * mangosd/realmd as Windows services support,
 * autogeneration mangosd/realmd crash reports (Windows only),
 * Visual Studio 2008 Express and Pro support,
 * division Mangos DataBase to MangosDB(WorldDB) and CharactersDB.
 * Upgrade to client version 2.2.3 (build 7359).

## Version 0.8
 * battleground support,
 * gender specific creatures,
 * game objects triggering more scripts,
 * more auras, effects, spells, and talents working,
 * dungeon specific data,
 * localization support for data,
 * vmaps for line of sight handling,
 * build support for OpenBSD and FreeBSD,
 * Upgrade to client version 2.1.3 (build 6898).

## Version 0.7, Codename "Eye of the Storm"
 * Full rewrite of the GNU autotools based build system, in order to
   smooth and clean up the build process.
 * Full rewrite of the Windows build system, allowing MaNGOS to be built
   from Visual C++ 2003 Toolkit, Visual C++ 2003, Visual C++ 2005.
 * Upgrade to client version 2.0.12 (build 6546).

## Version 0.6, Codename "Black Dragonflight"
 * MaNGOS development moved to http://sourceforge.net/projects/mangos/
 * A lots of speed improvements to data handling and data transfer have been
   completed.
 * Mail, auction house, profession, creature, and game object systems have seen
   a lot of improvements and can be considered feature complete. Some minor bugs
   still are left.
 * Rest system has been finished.
 * Initial pet stable support has been added.
 * PvP and duel system have seen lots of improvements
 * Guilds are not just a myth anymore. Coming closer to full guild support.
 * Taxi and transport system is close to a fully working system.
 * A lot more has been added.

## Version 0.5, Codename "Stable Master"
 * Core stability improvements, lots of threading and memory usage related
   bugs have been resolved.
 * MySQL Database backend has been cleaned up.
 * Proper support for game clients of version 1.10.2 has been added.
 * Support for auras, pets, spells, talents, totems, etc. has been added
   or improved.
 * Cross-platform issues have been resolved, MaNGOS should now build on
   more platforms, including FreeBSD.

## Version 0.1, Codename "Lightbringer"
 * AI system
 * Grid system
 * Lots a game features working.
 * Lovely optimizations for the beloved server.

## Version 0.0.3, Codename "Mango Carpet"
 * Interim release, not gone public.

## Version 0.0.2, Codename "Library"
 * Most data conversions working.
 * Database cleanup.
 * More features working.
 * Many bugs fixed.

## Version 0.0.1, Codename "Endeavour"
 * Data converter added.
 * Build system improvements.
 * Bug fixing and code cleaning.


src/realmd/Main.cpp
/*
 * This file is part of the CMaNGOS Project. See AUTHORS file for Copyright information
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <sstream>
#include <boost/date_time/posix_time/posix_time_duration.hpp>
#include <boost/program_options.hpp>
#include <boost/thread.hpp>
#include <boost/version.hpp>
#include <openssl/opensslv.h>
#include <openssl/crypto.h>
#include "AuthSocket.h"
#include "Common.h"
#include "Config/Config.h"
#include "Database/DatabaseEnv.h"
#include "Log.h"
#include "RealmList.h"
#include "revision.h"
#include "revision_nr.h"
#include "revision_sql.h"
#include "SessionManager.h"
#include "SystemConfig.h"
#include "Util.h"

#ifdef WIN32
#   include "ServiceWin32.h"
char serviceName[] = "realmd";
char serviceLongName[] = "MaNGOS realmd service";
char serviceDescription[] = "Massive Network Game Object Server";
/*
 * -1 - not in service mode
 *  0 - stopped
 *  1 - running
 *  2 - paused
 */
int serviceStatus = -1;
#else
#   include "PosixDaemon.h"
#endif

bool StartDatabase();

bool stopEvent = false;     // Setting it to true stops the server
DatabaseType LoginDatabase; // Accessor to the realm server database

/// Handle termination signals
// Put the global variable stopEvent to 'true' if a termination signal is caught
void HandleSignal(int signal)
{
    switch (signal)
    {
    case SIGINT:
    case SIGTERM:
#ifdef _WIN32
    case SIGBREAK:
#endif
        stopEvent = true;
        break;
    }
}

/// Print out the usage string for this program on the console.
void usage(boost::program_options::options_description const& desc, const char* prog)
{
    std::ostringstream ss;
    ss << desc;
    sLog.outString("Usage: \n %s [<options>]\n%s", prog, ss.str().c_str());
}

extern int main(int argc, char** argv)
{
    std::string cfg_file;
    std::string serviceDaemonMode;

    boost::program_options::options_description description("Allowed options");
    description.add_options()
        ("version,v", "print version and exit")
        ("help,h", "print commandline help and exit")
        ("config,c", boost::program_options::value<std::string>(&cfg_file)->default_value(_REALMD_CONFIG), "use as configuration file")
#ifdef WIN32
        ("service,s", boost::program_options::value<std::string>(&serviceDaemonMode), "running as service, arg functions: run, install, uninstall")
#else
        ("service,s", boost::program_options::value<std::string>(&serviceDaemonMode), "running as daemon, arg functions: run, stop")
#endif
        ;

    boost::program_options::variables_map vm;

    try {
        boost::program_options::store(boost::program_options::command_line_parser(argc, argv).
            options(description).run(), vm);
        boost::program_options::notify(vm);
    }
    catch(boost::program_options::unknown_option const& ex)
    {
        sLog.outError("Runtime-Error: unknown option %s", ex.get_option_name().c_str());
        usage(description, argv[0]);
        Log::WaitBeforeContinueIfNeed();
        return 1;
    }
    catch(boost::program_options::invalid_command_line_syntax const& ex)
    {
        sLog.outError("Runtime-Error: invalid syntax for option %s", ex.get_option_name().c_str());
        usage(description, argv[0]);
        Log::WaitBeforeContinueIfNeed();
        return 1;
    }

    if (vm.count("version"))
    {
        printf("%s\n", _FULLVERSION(REVISION_DATE, REVISION_TIME, REVISION_NR, REVISION_ID));
        return 0;
    }

    if (vm.count("help"))
    {
        usage(description, argv[0]);
        return 0;
    }

    if (!serviceDaemonMode.empty())
    {
#ifdef WIN32
        char const* const serviceModes[] =  { "run", "install", "uninstall", NULL };
#else
        char const* const serviceModes[] =  { "run", "stop", NULL };
#endif
        char const* const* mode_ptr = &serviceModes[0];
        for(; *mode_ptr != NULL; ++mode_ptr)
            if (*mode_ptr == serviceDaemonMode)
                break;

        if (!*mode_ptr)
        {
            sLog.outError("Runtime-Error: -s unsupported argument %s", serviceDaemonMode.c_str());
            usage(description, argv[0]);
            Log::WaitBeforeContinueIfNeed();
            return 1;
        }
    }

    // Windows service command need execute before config read
#ifdef WIN32
    switch (serviceDaemonMode[0])
    {
    case 'i':
        if (WinServiceInstall())
            sLog.outString("Installing service");
        return 1;
    case 'u':
        if (WinServiceUninstall())
            sLog.outString("Uninstalling service");
        return 1;
    case 'r':
        WinServiceRun();
        break;
    }
#endif

    if (!sConfig.SetSource(cfg_file, "RealmdConf"))
    {
        sLog.outError("Could not find configuration file %s.", cfg_file);
        Log::WaitBeforeContinueIfNeed();
        return 1;
    }

    // Posix daemon commands need apply after config read
#ifndef WIN32
    switch (serviceDaemonMode)
    {
        case 'r':
            startDaemon();
            break;
        case 's':
            stopDaemon();
            break;
    }
#endif

    sLog.Initialize();
    sLog.outString("%s [realm-daemon]", _FULLVERSION(REVISION_DATE, REVISION_TIME, REVISION_NR, REVISION_ID));
    sLog.outString("<Ctrl-C> to stop.\n");
    sLog.outString("Using configuration file %s.", cfg_file);

    // Check the version of the configuration file
    uint32 confVersion = sConfig.GetIntDefault("ConfVersion", 0);
    if (confVersion < _REALMDCONFVERSION)
    {
        sLog.outError("*****************************************************************************");
        sLog.outError(" WARNING: Your realmd.conf version indicates your conf file is out of date!");
        sLog.outError("          Please check for updates, as your current default values may cause");
        sLog.outError("          strange behavior.");
        sLog.outError("*****************************************************************************");
        Log::WaitBeforeContinueIfNeed();
    }

    DETAIL_LOG("%s (Library: %s)", OPENSSL_VERSION_TEXT, SSLeay_version(SSLEAY_VERSION));
    if (SSLeay() < 0x009080bfL)
    {
        DETAIL_LOG("WARNING: Outdated version of OpenSSL lib. Logins to server may not work!");
        DETAIL_LOG("WARNING: Minimal required version [OpenSSL 0.9.8k]");
    }
    DETAIL_LOG("Using BOOST: %i.%i.%i", BOOST_VERSION / 100000, BOOST_VERSION / 100 % 1000, BOOST_VERSION % 100);
    sLog.outBasic("Max allowed open files is %d", boost::asio::socket_base::max_connections);

    // Realmd PID file creation
    std::string pidfile = sConfig.GetStringDefault("PidFile", "");
    if (!pidfile.empty())
    {
        uint32 pid = CreatePIDFile(pidfile);
        if (!pid)
        {
            sLog.outError("Cannot create PID file %s.\n", pidfile.c_str());
            Log::WaitBeforeContinueIfNeed();
            return 1;
        }

        sLog.outString("Daemon PID: %u\n", pid);
    }

    // Initialize the database connection
    if (!StartDatabase())
    {
        Log::WaitBeforeContinueIfNeed();
        return 1;
    }

    // Get the list of realms for the server
    sRealmList.Initialize(sConfig.GetIntDefault("RealmsStateUpdateDelay", 20));
    if (sRealmList.size() == 0)
    {
        sLog.outError("No valid realms specified.");
        Log::WaitBeforeContinueIfNeed();
        return 1;
    }

    // Cleanup query
    // Set expired bans to inactive
    LoginDatabase.BeginTransaction();
    LoginDatabase.Execute("UPDATE account_banned SET active = 0 WHERE unbandate<=UNIX_TIMESTAMP() AND unbandate<>bandate");
    LoginDatabase.Execute("DELETE FROM ip_banned WHERE unbandate<=UNIX_TIMESTAMP() AND unbandate<>bandate");
    LoginDatabase.CommitTransaction();

    // Launch the listening network socket
    uint16 rmport = sConfig.GetIntDefault("RealmServerPort", DEFAULT_REALMSERVER_PORT);
    std::string bind_ip = sConfig.GetStringDefault("BindIP", "0.0.0.0");

    std::auto_ptr<SessionManager> manager(new SessionManager());
    if (!manager->StartNetwork(rmport, bind_ip))
    {
        sLog.outError("MaNGOS realmd can not bind to %s:%d", bind_ip.c_str(), rmport);
        Log::WaitBeforeContinueIfNeed();
        return 1;
    }

    // Catch termination signals
    signal(SIGINT, HandleSignal);
    signal(SIGTERM, HandleSignal);
#ifdef _WIN32
    signal(SIGBREAK, HandleSignal);
#endif

    // Handle affinity for multiple processors and process priority on Windows
#ifdef WIN32
    {
        HANDLE hProcess = GetCurrentProcess();

        uint32 Aff = sConfig.GetIntDefault("UseProcessors", 0);
        if (Aff > 0)
        {
            ULONG_PTR appAff;
            ULONG_PTR sysAff;

            if (GetProcessAffinityMask(hProcess, &appAff, &sysAff))
            {
                ULONG_PTR curAff = Aff & appAff; // Remove non accessible processors

                if (!curAff)
                    sLog.outError("Processors marked in UseProcessors bitmask (hex) %x not accessible for realmd. Accessible processors bitmask (hex): %x", Aff, appAff);
                else
                {
                    if (SetProcessAffinityMask(hProcess, curAff))
                        sLog.outString("Using processors (bitmask, hex): %x", curAff);
                    else
                        sLog.outError("Can't set used processors (hex): %x", curAff);
                }
            }
            sLog.outString();
        }

        bool prio = sConfig.GetBoolDefault("ProcessPriority", false);

        if (prio)
        {
            if (SetPriorityClass(hProcess, HIGH_PRIORITY_CLASS))
                sLog.outString("realmd process priority class set to HIGH");
            else
                sLog.outError("Can't set realmd process priority class.");
            sLog.outString();
        }
    }
#endif

    // Server has started up successfully => enable async DB requests
    LoginDatabase.AllowAsyncTransactions();

    // Maximum counter for next ping
    uint32 numLoops = (sConfig.GetIntDefault("MaxPingTime", 30) * (MINUTE * 1000000 / 100000));
    uint32 loopCounter = 0;

#ifndef WIN32
    detachDaemon();
#endif

    // Wait for termination signal
    while (!stopEvent)
    {
        // Dont move this outside the loop, the reactor will modify it
        boost::this_thread::sleep(boost::posix_time::milliseconds(10));

        if ((++loopCounter) == numLoops)
        {
            loopCounter = 0;
            DETAIL_LOG("Ping MySQL to keep connection alive");
            LoginDatabase.Ping();
        }

#ifdef WIN32
        if (serviceStatus == 0)
            stopEvent = true;

        while (serviceStatus == 2)
            Sleep(1000);
#endif
    }

    manager.reset();

    // Wait for the delay thread to exit
    LoginDatabase.HaltDelayThread();

    sLog.outString("Halting process...");
    return 0;
}

/// Initialize connection to the database
bool StartDatabase()
{
    std::string dbstring = sConfig.GetStringDefault("LoginDatabaseInfo", "");
    if (dbstring.empty())
    {
        sLog.outError("Database not specified");
        return false;
    }

    sLog.outString("Login Database total connections: %i", 1 + 1);

    if (!LoginDatabase.Initialize(dbstring.c_str()))
    {
        sLog.outError("Cannot connect to database");
        return false;
    }

    if (!LoginDatabase.CheckRequiredField("realmd_db_version", REVISION_DB_REALMD))
    {
        // Wait for already started DB delay threads to end
        LoginDatabase.HaltDelayThread();
        return false;
    }

    return true;
}